apis/stacks/v1alpha1/stackdefinition.go

```diff
@@ -0,0 +1,120 @@
+/*
+Copyright 2019 The Crossplane Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+)
+
+// NOTE: json tags are required.  Any new fields you add must have json tags for the fields to be serialized.
+
+// StackDefinitionSpec defines the desired state of StackDefinition
+type StackDefinitionSpec struct {
+	AppMetadataSpec `json:",inline"`
+	CRDs            CRDList                 `json:"customresourcedefinitions,omitempty"`
+	Controller      ControllerSpec          `json:"controller,omitempty"`
+	Permissions     PermissionsSpec         `json:"permissions,omitempty"`
+	Behavior        StackDefinitionBehavior `json:"behavior,omitempty"`
+}
+
+// StackDefinitionSource is the stack image which this stack configuration is from.
+// In the future, other source types may be supported, such as a URL.
+type StackDefinitionSource struct {
+	// a container image id
+	Image string `json:"image,omitempty"`
+}
+
+// StackDefinitionBehavior specifies the behavior for the stack, if the stack has behaviors instead of a controller
+type StackDefinitionBehavior struct {
+	CRD    StackDefinitionBehaviorCRD       `json:"crd,omitempty"`
+	Engine StackResourceEngineConfiguration `json:"engine,omitempty"`
+	// Theoretically, source and engine could be specified at a per-hook level
+	// as well.
+	Source    StackDefinitionSource            `json:"source,omitempty"`
+	Reconcile StackDefinitionHookConfiguration `json:"reconcile,omitempty"`
+}
+
+// StackDefinitionBehaviorCRD represents the CRD which the stack's behavior controller will watch. When CRs of this
+// CRD kind appear and are modified in the cluster, the behavior will execute.
+type StackDefinitionBehaviorCRD struct {
+	APIVersion string `json:"apiVersion"`
+	Kind       string `json:"kind"`
+}
+
+// StackResourceEngineConfiguration represents a configuration for a resource engine, such as helm2 or kustomize.
+type StackResourceEngineConfiguration struct {
+	// Type is the engine type, such as "helm2" or "kustomize"
+	Type string `json:"type"`
+	// Because different engine configurations could specify conflicting field names,
+	// we want to namespace the engines with engine-specific keys
+	Kustomize *KustomizeEngineConfiguration `json:"kustomize,omitempty"`
```

Comment:
I originally was going to inline all the fields, but I decided against it in case different engines have fields with the same names. Inline they'd conflict, but they don't if we put them under an engine-specific field name.

Reply:
Got it. Then maybe we can omit `Type`? [`VolumeSource`](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/api/core/v1/types.go#L51) doesn't have a field to specify its type and in the implementation, `switch` is used like `if kustomize is not nil then this is kustomize or if helm is not nil this is helm etc`